core/primitives-core/src/account.rs

```diff
@@ -20,6 +23,173 @@ pub struct Account {
+#[derive(BorshSerialize, BorshDeserialize, Serialize, Deserialize, PartialEq, Eq, Debug, Clone)]
+pub struct AccountV1 {
+    /// The total not locked tokens.
+    #[serde(with = "u128_dec_format_compatible")]
+    pub amount: Balance,
+    /// The amount locked due to staking.
+    #[serde(with = "u128_dec_format_compatible")]
+    pub locked: Balance,
+    /// Hash of the code stored in the storage for this account.
+    pub code_hash: CryptoHash,
+    /// Storage used by the given account, includes account id, this struct, access keys and other data.
+    pub storage_usage: StorageUsage,
+}
+
+#[cfg(feature = "protocol_feature_add_account_versions")]
+#[derive(BorshSerialize, Serialize, Deserialize, PartialEq, Eq, Debug, Clone)]
+pub enum Account {
+    AccountV1(AccountV1),
+}
+
+impl Account {
+    #[cfg(not(feature = "protocol_feature_add_account_versions"))]
+    pub fn new(
+        amount: Balance,
+        locked: Balance,
+        code_hash: CryptoHash,
+        storage_usage: StorageUsage,
+    ) -> Self {
+        Account { amount, locked, code_hash, storage_usage }
+    }
+
+    #[cfg(feature = "protocol_feature_add_account_versions")]
+    pub fn new(
+        amount: Balance,
+        locked: Balance,
+        code_hash: CryptoHash,
+        storage_usage: StorageUsage,
+    ) -> Self {
+        Account::AccountV1(AccountV1 { amount, locked, code_hash, storage_usage })
+    }
+
+    #[cfg(not(feature = "protocol_feature_add_account_versions"))]
+    pub fn amount(&self) -> Balance {
+        self.amount
+    }
+
+    #[cfg(feature = "protocol_feature_add_account_versions")]
+    pub fn amount(&self) -> Balance {
+        match self {
+            Account::AccountV1(acc) => acc.amount,
+        }
+    }
+
+    #[cfg(not(feature = "protocol_feature_add_account_versions"))]
+    pub fn locked(&self) -> Balance {
+        self.locked
+    }
+
+    #[cfg(feature = "protocol_feature_add_account_versions")]
+    pub fn locked(&self) -> Balance {
+        match self {
+            Account::AccountV1(acc) => acc.locked,
+        }
+    }
+
+    #[cfg(not(feature = "protocol_feature_add_account_versions"))]
+    pub fn code_hash(&self) -> CryptoHash {
+        self.code_hash
+    }
+
+    #[cfg(feature = "protocol_feature_add_account_versions")]
+    pub fn code_hash(&self) -> CryptoHash {
+        match self {
+            Account::AccountV1(acc) => acc.code_hash,
+        }
+    }
+
+    #[cfg(not(feature = "protocol_feature_add_account_versions"))]
+    pub fn storage_usage(&self) -> StorageUsage {
+        self.storage_usage
+    }
+
+    #[cfg(feature = "protocol_feature_add_account_versions")]
+    pub fn storage_usage(&self) -> StorageUsage {
+        match self {
+            Account::AccountV1(acc) => acc.storage_usage,
+        }
+    }
+
+    #[cfg(not(feature = "protocol_feature_add_account_versions"))]
+    pub fn set_amount(&mut self, amount: Balance) {
+        self.amount = amount;
+    }
```

Comment:
Given that we introduce functions here anyway, perhaps we can introduce 

```
pub fn inc_amount(&mut self, by: Balance) -> Result<(), OverflowError> {
    self.amount = self.amount.checked_add(by).ok_or(())?;
    Ok());
}
pub fn dec_amount(&mut self, by: Balance) -> Result<(), OverflowError>
``` 

instead? That way, "not forget overflow checks" becomes less of an issue. 

They do seem to help in the common case:

```diff
diff --git a/core/primitives-core/src/account.rs b/core/primitives-core/src/account.rs
index 2d5442fa..909fdd06 100644
--- a/core/primitives-core/src/account.rs
+++ b/core/primitives-core/src/account.rs
@@ -113,9 +113,15 @@ impl Account {
         }
     }
 
+    // #[cfg(not(feature = "protocol_feature_add_account_versions"))]
+    // pub fn set_amount(&mut self, amount: Balance) {
+    //     self.amount = amount;
+    // }
+
     #[cfg(not(feature = "protocol_feature_add_account_versions"))]
-    pub fn set_amount(&mut self, amount: Balance) {
-        self.amount = amount;
+    pub fn inc_amount(&mut self, by: Balance) -> Result<(), ()> {
+        self.amount = self.amount.checked_add(by).ok_or(())?;
+        Ok(())
     }
 
     #[cfg(feature = "protocol_feature_add_account_versions")]
diff --git a/runtime/runtime/src/actions.rs b/runtime/runtime/src/actions.rs
index 8338a56a..3d77b090 100644
--- a/runtime/runtime/src/actions.rs
+++ b/runtime/runtime/src/actions.rs
@@ -329,9 +329,9 @@ pub(crate) fn action_transfer(
     account: &mut Account,
     transfer: &TransferAction,
 ) -> Result<(), StorageError> {
-    account.set_amount(account.amount().checked_add(transfer.deposit).ok_or_else(|| {
+    account.inc_amount(transfer.deposit).map_err(|()| {
         StorageError::StorageInconsistentState("Account balance integer overflow".to_string())
-    })?);
+    })?;
     Ok(())
 }
 
diff --git a/runtime/runtime/src/lib.rs b/runtime/runtime/src/lib.rs
index 1e466789..c06b9bca 100644
--- a/runtime/runtime/src/lib.rs
+++ b/runtime/runtime/src/lib.rs
@@ -7,6 +7,7 @@ use log::debug;
 pub use near_crypto;
 use near_crypto::PublicKey;
 pub use near_primitives;
+use near_primitives::errors::IntegerOverflowError;
 use near_primitives::runtime::get_insufficient_storage_stake;
 use near_primitives::{
     account::{AccessKey, Account},
@@ -604,7 +605,7 @@ impl Runtime {
                 // account holder. If the account doesn't exist by the end of the execution, the
                 // validators receive the full reward.
                 tx_burnt_amount -= receiver_reward;
-                account.set_amount(safe_add_balance(account.amount(), receiver_reward)?);
+                account.inc_amount(receiver_reward).map_err(|()| IntegerOverflowError)?;
                 set_account(state_update, account_id.clone(), account);
                 state_update.commit(StateChangeCause::ActionReceiptGasReward {
                     receipt_hash: receipt.get_hash(),
@@ -984,12 +985,9 @@ impl Runtime {
                         .checked_sub(return_stake)
                         .ok_or_else(|| RuntimeError::UnexpectedIntegerOverflow)?,
                 );
-                account.set_amount(
-                    account
-                        .amount()
-                        .checked_add(return_stake)
-                        .ok_or_else(|| RuntimeError::UnexpectedIntegerOverflow)?,
-                );
+                account
+                    .inc_amount(return_stake)
+                    .map_err(|()| RuntimeError::UnexpectedIntegerOverflow);
 
                 set_account(state_update, account_id.clone(), &account);
             } else if *max_of_stakes > 0 {
@@ -1050,12 +1048,9 @@ impl Runtime {
                             account_id
                         ))
                     })?;
-                account.set_amount(
-                    account
-                        .amount()
-                        .checked_add(treasury_reward)
-                        .ok_or_else(|| RuntimeError::UnexpectedIntegerOverflow)?,
-                );
+                account
+                    .inc_amount(treasury_reward)
+                    .map_err(|()| RuntimeError::UnexpectedIntegerOverflow)?;
                 set_account(state_update, account_id.clone(), &account);
             }
         }
```

Though, there seems to be the case where we genuinely need `set` sadly. 

But this refactor is probably not related to this PR and can be done separatly.